Tests/Utils/Builders/ConversationEngineTestBuilder.php

```diff
@@ -0,0 +1,398 @@
+
+        $this->testCase = $testCase;
+    }
+
+    /**
+     *  Dehydrate the conversation object (Set all properties to null)
+     *
+     * @return $this
+     */
+    private function dehydrate(): ConversationEngineTestBuilder
+    {
+        $this->scenario = null;
+        $this->states = null;
+        $this->utterance = null;
+        $this->user = null;
+        $this->interpretersToInterpretedIntents = null;
+        $this->messages = null;
+        $this->userState = null;
+        $this->questionIntentOdId = null;
+        $this->answerIntentOdId = null;
+
+        return $this;
+    }
+
+    /**
+     * Resetting Conversation Engine for cleaning chat history
+     *
+     * @return $this
+     */
+    private function renewConversationEngine(): ConversationEngineTestBuilder
+    {
+        $this->conversationEngine = resolve(ConversationEngineInterface::class);
+
+        return $this;
+    }
+
+    /**
+     * Setting up scenario from path string
+     *
+     * @param string $path
+     * @return $this
+     */
+    private function setScenario(string $path): ConversationEngineTestBuilder
+    {
+        $scenarioFileData = File::get($path);
+        $this->scenario = ScenarioImportExportHelper::importScenarioFromString($scenarioFileData);
+
+        return $this;
+    }
+
+    /**
+     * Setting up listeners for update invents
+     *
+     * @return $this
+     */
+    private function seUpListeners(): ConversationEngineTestBuilder
+    {
+        $this->states = new Map();
+
+        resolve(Dispatcher::class)->listen(IncomingIntentStateUpdate::class, function (IncomingIntentStateUpdate $event) {
+            $this->states->put(get_class($event), $event->getConversationalState());
+        });
+
+        resolve(Dispatcher::class)->listen(OutgoingIntentStateUpdate::class, function (OutgoingIntentStateUpdate $event) {
+            $this->states->put(get_class($event), $event->getConversationalState());
+        });
+
+        resolve(Dispatcher::class)->listen(MessageTemplateStateUpdate::class, function (MessageTemplateStateUpdate $event) {
+            $this->states->put(get_class($event), $event->getConversationalState());
+        });
+
+        return $this;
+    }
+
+    /**
+     * Preparing scenario
+     *
+     * @param string $path
+     * @return $this
+     */
+    public function initializeScenario(string $path): ConversationEngineTestBuilder
+    {
+        $this->dehydrate()
+            ->renewConversationEngine()
+            ->setScenario($path)
+            ->seUpListeners();
+
+        ContextService::saveAttribute(
+            UserContext::USER_CONTEXT . '.' . 'selected_scenario',
+            $this->scenario->getUid()
+        );
+
+        return $this;
+    }
+
+    /**
+     * Set up Utterance (what user said)
+     *
+     * @param string|UtteranceAttribute $utteranceOrText
+     * @return $this
+     */
+    public function userSaid($utteranceOrText): ConversationEngineTestBuilder
```

Comment:
What do you think about how to name this method? I'm wondering if they'd be best in present tense, like follows.

```suggestion
    public function userSays($utteranceOrText): ConversationEngineTestBuilder
```